src/libxtp/bulkesp.cc

```diff
@@ -0,0 +1,768 @@
+                }
+                
+                system_netcharge+=netcharge;
+
+
+                //and save it to a .cube file
+                std::ostringstream fn;
+                if (periodic && _grid.getCubegrid()) {
+                    fn.str(std::string());
+                    fn << "BulkEsp_" << m - mols.begin() << ".cube";
+                    _grid.printgridtoCubefile(fn.str());
+                }
+
+                //output
+                //CHELPG grids aren't periodic and equally spaced,
+                //so can't output to .cube format.
+                //Create own format.
+                //note: just like printgridtoCubefile, this prints potential from apolar sites
+                fn.clear();
+                fn.str("");
+                fn << "BulkEsp_" << m - mols.begin() << ".grid";
+                _grid.writeIrregularGrid(fn.str(), m->atoms, _ECP);
+
+
+                //TODO: fit charges
+                std::vector< tools::vec > _fitcenters;
+
+                for (unsigned j = 0; j < m->atoms.size(); j++) {
+                    tools::vec _pos = m->atoms[j]->getPos() * tools::conv::ang2nm;
+                    _fitcenters.push_back(_pos);
+                }
+
+                std::vector<double> _charges = FitPartialCharges(_fitcenters, _grid, ESP, netcharge);
+
+
+                //Write charges to qmatoms
+                for (unsigned _i = 0; _i < m->atoms.size(); _i++) {
+                    m->atoms[_i]->charge = _charges[_i];
+                }
+
+
+                CTP_LOG(ctp::logDEBUG, *_log) << "Bulkesp::Evaluate(): " << ctp::TimeStamp() << " done with molecule " << m - mols.begin() << endl << flush;
+                //exit(-1);
+
+            }
+            CTP_LOG(ctp::logDEBUG, *_log) << "Bulkesp::Evaluate(): " << ctp::TimeStamp() << " All molecules processed." << endl << flush;
+            dipolesLog->close();
+
+            CTP_LOG(ctp::logDEBUG, *_log) << "Bulkesp::Evaluate(): " << ctp::TimeStamp() << " Net charge of the whole system is "<< system_netcharge << endl << flush;
+            if(fabs(system_netcharge)>0.05){
+                if(fmod(system_netcharge,1.0)<0.05){
+                    CTP_LOG(ctp::logWARNING, *_log) << "Bulkesp::Evaluate(): " << ctp::TimeStamp() << " System is ionized. Is this intended?"<< endl << flush;
+                }
+                else{
+                    CTP_LOG(ctp::logWARNING, *_log) << "Bulkesp::Evaluate(): " << ctp::TimeStamp() << " System has significant net partial charge. Projection was likely incomplete. "<< endl << flush;
+                }
+            }
+        }
+
+        
+        
+        
+        
+        /*
+         * Modified version of espfit::FitPartialCharges() that allows for periodic boundary conditions
+         * 
+         */
+        std::vector<double> Bulkesp::FitPartialCharges(std::vector< tools::vec >& _fitcenters, Grid& _grid, ub::vector<double>& _potential, double& _netcharge) {
```

Comment:
use const declarations for input arguments

Reply:
This is nearly a complete copy of FitPartialCharges. Let us talk about refactoring ESP to reduce code duplication.